### ISSUE TITLE
Finding Jump In nodes

### ISSUE THREAD

**Issue by kenfhill84083** [freeplane/freeplane#1213]
**Is your feature request related to a problem? Please describe.**
Sometimes I have wished for a way to identify and find Jump In nodes.  I know I have Jumped into a node but I cannot remember much about the content of the node or children.

**Request**
I would like to be able to use the Filter toolbar - or whatever functionality might be better - to quickly find all the Jump In nodes in a map.

**background**
I have created a Jump In style I can filter on - but it involves me consistently applying the style.  And I still have Jump In nodes I made prior to this style.

**Comment by macmarrum** [comment]
I know that Jump In is a view-related feature, i.e. it isn't saved with the mindmap. The chain of Jump-Ins is remembered only until Freeplane is closed.

I can think of a solution that uses a script (aka macro) for jumping in. The script would apply a custom style and then do the jump-in. One would use the script instead of the Jump In action.

The script (as any script) can be bound to a hot key.

Here's the script
```groovy
// @ExecutionModes({ON_SINGLE_NODE})

// put the name of your custom style, if different
def jumpInStyleName = 'Jump In'

def styleName = node.style.name
if (!styleName) {
    // no style -- apply jumpInStyleName
    node.style.name = jumpInStyleName
} else if (styleName == jumpInStyleName) {
    // jumpInStyleName is already applied
} else {
    // there is already a style applied, and it isn't jumpInStyleName
    // save the currently applied style as a Node Conditional Style, if not already there
    def styleAlreadyAdded = node.conditionalStyles.collect().find { it.styleName == styleName }
    if (!styleAlreadyAdded)
        node.conditionalStyles.add(true, null, styleName, false)
    // apply jumpInStyleName
    node.style.name = jumpInStyleName
}
menuUtils.executeMenuItems(['JumpInAction'])
```

**Comment by kenfhill84083** [comment]
> I know that Jump In is a view-related feature, i.e. it isn't saved with the mindmap. The chain of Jump-Ins is remembered only until Freeplane is closed.
>
> I can think of a solution that uses a script (aka macro) for jumping in. The script would apply a custom style and then do the jump-in. One would use the script instead of the Jump In action.
>
>The script (as any script) can be bound to a hot key.

Works great.  Thanks.  Looks like that is as close as I'll get with my request so I'll close it out.

**Comment by EdoFro** [comment]
> I know that Jump In is a view-related feature, i.e. it isn't saved with the mindmap. The chain of Jump-Ins is remembered only until Freeplane is closed.

I think one way to discover a "Once Jumped In" node is by looking at the `sideAtRoot` property of its children.

Here are two different attempts:

### Test 1
```groovy
node.mindMap.root.find{n -> !n.root && n.children.any{it.sideAtRoot.toString() in ['BOTTOM_OR_RIGHT', 'TOP_OR_LEFT']}}
```

### Test 2
```groovy
node.mindMap.root.find{n ->
    def branchNodeSide = n.pathToRoot.drop(1)[0]?.sideAtRoot
    return !n.root && n.children.any{m -> m.sideAtRoot != branchNodeSide && m.sideAtRoot != org.freeplane.api.Side.DEFAULT}
}

```

**Comment by dpolivaev** [comment]
There is no such thing as jump-in nodes. I do not understand how you mean it and why you want to track it. Could you please elaborate?

**Comment by kenfhill84083** [comment]
> > I know that Jump In is a view-related feature, i.e. it isn't saved with the mindmap. The chain of Jump-Ins is remembered only until Freeplane is closed.
> 
> I think one way to discover a "Once Jumped In" node is by looking at the `sideAtRoot` property of its children.

Thanks.  Maybe someday I will get into scripting.  If that ever happens I will try to remember this thread.

Does this property mean I would need to have moved nodes to the left side of the temporary root node for this to work?

**Comment by kenfhill84083** [comment]
> There is no such thing as jump-in nodes. I do not understand how you mean it and why you want to track it. Could you please elaborate?

_jump-in nodes_ : It is my self-invented terminology.  I mean any node where I invoked Jump In.

A few times I have caught myself thinking, "Where is that node where I invoked Jump In and then moved those nodes to the left-hand side of a temporary root node?"  Then I caught myself thinking, "I wish I could find all those nodes where I invoked Jump In - so I could track down the node I'm thinking about (in my initial wondering)".

It is not a super high priority for me but it would be handy.

**Comment by dpolivaev** [comment]
Moving nodes to another side on the jump-in root has no implication on the side where they appear after jump-out unless you explicitly enable both-sides - layout for this node.

**Comment by dpolivaev** [comment]
So I do not understand why you might be interested in "tracking".
Also, you can have multiple views with different jump-in roots.

**Comment by kenfhill84083** [comment]
> So I do not understand why you might be interested in "tracking". Also, you can have multiple views with different jump-in roots.

_tracking_  I just mean finding such nodes.   The only reason I want to find them is I can't remember enough detail about what is in the node to effectively search for it - except that I remember it was in a node where I invoked Jump In.

Perhaps since it seems difficult for me to explain, the problem is just my old brain.

I am OK to just close this request.  My style and the script provided by @macmarrum are functioning fine as a work-around.

**Comment by macmarrum** [comment]
It's clear to me why such "tracking" might be of use.

In most of my mindmaps, I have only several main nodes that can act as View Root. But in some mindmaps, I have many potential nodes for a Jump-In. Marking them in some way, e.g. by applying a style (like **Jump In**) makes it easier to come back to them e.g. after a week.

**Comment by dpolivaev** [comment]
You can jump-in at any node.
If you want to create special styles for nodes you believe you want to jump-in you can do it, but it has no relevance for Freeplane.

**Comment by kenfhill84083** [comment]
> You can jump-in at any node. If you want to create special styles for nodes you believe you want to jump-in you can do it, but it has no relevance for Freeplane.

I don't want to apply a style to a node where I may want to jump in.  I apply the style to nodes where I already invoked jump in - so I can find them later.  Without the style I cannot find them because they look like every other node.

**Comment by macmarrum** [comment]
My Feature Request might be of some help in managing/recalling Jump-In nodes after a while:
- #1210

**Comment by dpolivaev** [comment]
What makes such nodes special after you jump in?

**Comment by macmarrum** [comment]
Their children are at a different side then before Jump-In

**Comment by dpolivaev** [comment]
No. They should not.

**Comment by macmarrum** [comment]
Here's an example.

Initially, `AAA` and `BBB` have all their children on one side.
After you jump in `AAA`, `a-1` switches to **left**.
Similarly for `BBB`.

[FP1213-JumpIn-tracking.xi.mm.zip](https://github.com/freeplane/freeplane/files/11678686/FP1213-JumpIn-tracking.xi.mm.zip)

**Comment by dpolivaev** [comment]
Yes, if you jump in you can distribute nodes at both sides. It has no impact after jump-out unless the node is configured for two sides layout. So, it doesn't have any implications after jump-out.

**Comment by macmarrum** [comment]
So you see that if I forget I jumped in `AAA` (let's say a week ago) and moved `a-1` from the (automatic) **right** to **left**, it makes sense to mark `AAA` somehow to help me remember?

**Comment by dpolivaev** [comment]
Why? It has effect only if you jump in again

**Comment by macmarrum** [comment]
Yes. I want to jump in again. And since I cannot tell which node was a View Root a week ago just by looking at it, I would need to search for it by brute force: jumping in all possible nodes.

I just need help to recall which node it was.

**Comment by dpolivaev** [comment]
Why might you be interested in knowing it?

**Comment by macmarrum** [comment]
I'm not sure I can explain any further.

Please note that my goal here isn't to advocate for any changes in Freeplane in relation to this Feature Request.

I am happy with where we are with the discussion. The topic has been thoroughly analyzed. Many good questions were asked and people expressed their opinions, provided ideas, etc.

I feel our time can be spent better elsewhere.

**Comment by amariusz** [comment]
I think the OP question could be rephrased to "finding nodes which would have children distributed differently after Jumping in".

> Why might you be interested in knowing it?

@kenfhill84083 has published several nice videos about this feature that in my opinion explains quite well the reasoning behind this question. For example:
- use some nodes as minimaps  - https://youtu.be/CJm-19pxQxo?t=574
- workaround for editing long nodes on the left side - https://youtu.be/2r6HPw11UCg?t=100

In my opinion having the ability to _find_ all such nodes and then apply an icon/style could be useful sometimes.
@EdoFro approach with a small script also seems like a good solution.

~~Edit: I've just realized that after closing mindmap the change in children distrbution made within Jump In mode is lost. As @macmarrum stated it's a 'view-related feature'. So there's no benefit in implementing such find capability.~~

Edit: I've checked it again on 1.11.5-pre14 and observed that the distribution is preserved after reopening the file. I'm not sure why it wasn't last time - maybe I've selected non-default children layout, run into some bug or just messed up.
